### PR TITLE
Rewrite `workshops/urls.py` to nest URLs under common hierarchy.

### LINKED ISSUE #987: Fails to return 404 at specific URLs
`workshops/person/example` should return 404, returns 500 instead.

Fix:
```git
-    url(r'^person/(?P<person_id>[\w\.-]+)/?$', views.person_details, name='person_details'),
+    url(r'^person/(?P<person_id>\d+)/?$', views.person_details, name='person_details'),
```

### PR DESCRIPTION
**Pull request by narayanaditya95**
Fixes #987

@pbanaszkiewicz Do you approve of the refactor?

### CODE REVIEW

File: workshops/urls.py [hunk 1]
```diff
@@ -25,18 +25,22 @@
...
-    url(r'^persons/merge/$', views.persons_merge, name='persons_merge'),
-    url(r'^persons/sync_usersocialauth/(?P<person_id>\d+)/?', views.sync_usersocialauth, name='sync_usersocialauth'),
+    url(r'^persons/', include([
+        url(r'^$', views.all_persons, name='all_persons'),
+        url(r'^add$', views.PersonCreate.as_view(), name='person_add'),
+        url(r'^bulkadd/$',views.person_bulk_add, name='person_bulk_add'),
+        url(r'^bulkadd/template',views.person_bulk_add_template, name='person_bulk_add_template'),
+        url(r'^bulkadd/confirm$',views.person_bulk_add_confirmation, name='person_bulk_add_confirmation'),
+        url(r'^merge/$', views.persons_merge, name='persons_merge'),
+        url(r'^sync_usersocialauth/(?P<person_id>\d+)/?', views.sync_usersocialauth, name='sync_usersocialauth'),
```
Comment (pbanaszkiewicz) [inline workshops/urls.py]: Maybe change this URL to be in the other group?

**Comment by pbanaszkiewicz** [comment]
I approve, it looks nicer visually, thank you. Ping me back when you move the URL I commented about.

File: workshops/urls.py (hunk 1 above)
Comment (narayanaditya95) [inline workshops/urls.py]: Yes, that would be better.

**Comment by narayanaditya95** [comment]
@pbanaszkiewicz done.

**Comment by narayanaditya95** [comment]
@pbanaszkiewicz Do you want other URLs to be refactored?

**Comment by pbanaszkiewicz** [comment]
@narayanaditya95 yes please and thank you very much :-)

**Comment by narayanaditya95** [comment]
Last question, I see that we sometimes mandate that URLs end with `/` and sometimes we don't.
What should be the convention? Or do I maintain the status quo?

```py
    url(r'^admin-dashboard/$', views.admin_dashboard, name='admin-dashboard'), # enforce /
    url(r'^events/?$', views.all_events, name='all_events'), #optional
    url(r'^event/(?P<slug>[\w-]+)/?$', views.event_details, name='event_details'), #optional
    url(r'^event/(?P<slug>[\w-]+)/assign$', views.event_assign, name='event_assign'), # no /
```

**Comment by pbanaszkiewicz** [comment]
@narayanaditya95 let's use explicit `/` everywhere - and make that a separate commit change so that we revert easily if something happens.

**Comment by narayanaditya95** [comment]
@pbanaszkiewicz ping.

File: workshops/urls.py [hunk 2]
```diff
@@ -9,68 +9,94 @@
...
-    url(r'^invoice/(?P<request_id>\d+)/edit$', views.InvoiceRequestUpdate.as_view(), name='invoicerequest_edit'),
+    url(r'^organizations/', include([
+        url(r'^$', views.all_organizations, name='all_organizations'),
+        url(r'^add/$', views.OrganizationCreate.as_view(), name='organization_add'),
+    ])),
+    url(r'^organization/(?P<org_domain>[\w\.-]+)/', include([
+        url(r'^$', views.organization_details, name='organization_details'),
+        url(r'^edit$', views.OrganizationUpdate.as_view(), name='organization_edit'),
+        url(r'^delete$', views.OrganizationDelete.as_view(), name='organization_delete'),
+        url(r'membership/add', views.membership_create, name='membership_add'),
```
Comment (pbanaszkiewicz) [inline workshops/urls.py]: Should this start with `^` as other URLs?

File: workshops/urls.py [hunk 3]
```diff
@@ -176,22 +176,22 @@
...
         url(r'^$', views.trainingrequest_details, name='trainingrequest_details'),
-        url(r'^edit/?$', views.TrainingRequestUpdate.as_view(), name='trainingrequest_edit'),
+        url(r'^edit/$', views.TrainingRequestUpdate.as_view(), name='trainingrequest_edit'),
     ])),
 
-    url(r'^trainees/?$', views.all_trainees, name='all_trainees'),
+    url(r'^trainees/$', views.all_trainees, name='all_trainees'),
 
-    url(r'^training_progresses/add/(?P<trainee_id>\d+)?$', views.TrainingProgressCreate.as_view(), name='trainingprogress_add'),
+    url(r'^training_progresses/add/(?P<trainee_id>\d+)?/$', views.TrainingProgressCreate.as_view(), name='trainingprogress_add'),
```
Comment (pbanaszkiewicz) [inline workshops/urls.py]: Not sure how we should treat this URL. The group part `(?P<trainee_id>\d+)` seems to be optional, so with enforced trailing slash we theoretically could get `/workshops/training_progresses/add//` as a correct URL.

Comment (narayanaditya95) [inline workshops/urls.py]: Nice catch

File: workshops/urls.py [hunk 4]
```diff
@@ -61,21 +61,21 @@
...
         url(r'^$', views.event_details, name='event_details'),
-        url(r'^assign$', views.event_assign, name='event_assign'),
-        url(r'^assign/(?P<person_id>[\w\.-]+)$', views.event_assign, name='event_assign'),
-        url(r'^edit$', views.event_edit, name='event_edit'),
-        url(r'^delete$', views.EventDelete.as_view(), name='event_delete'),
+        url(r'^assign/$', views.event_assign, name='event_assign'),
+        url(r'^assign/(?P<person_id>[\w\.-]+)/$', views.event_assign, name='event_assign'),
+        url(r'^edit/$', views.event_edit, name='event_edit'),
+        url(r'^delete/$', views.EventDelete.as_view(), name='event_delete'),
         url(r'^validate/?$', views.validate_event, name='validate_event'),
```
Comment (pbanaszkiewicz) [inline workshops/urls.py]: Optional trailing slash

File: workshops/urls.py (hunk 3 above)
Comment (narayanaditya95) [inline workshops/urls.py]: `'^training_progresses/add/((?P<trainee_id>\d+)/)?$'`
This seems to work.

Comment (pbanaszkiewicz) [inline workshops/urls.py]: :+1:

**Comment by narayanaditya95** [comment]
@pbanaszkiewicz I ended up splitting the URL in two. I hope that's fine.

**Comment by pbanaszkiewicz** [comment]
@narayanaditya95 yes, thank you.

**Pull request merged**